### PR TITLE
Improved blocking

### PR DESCRIPTION
**Pull request by jadr2ddude**
This adds buffered channels, changing the channel implementation as little as possible.

### CODE REVIEW

**Comment by jadr2ddude** [comment]
fixes #510

**Comment by jadr2ddude** [comment]
After some discussion, we decided that we should refactor how we handle blocking before we add any more complexity to this.

**Comment by jadr2ddude** [comment]
The improved blocking has been added

**Comment by aykevl** [comment]
@jadr2ddude can you rebase on the dev branch? There are some conflicts.
Also, I just enabled building pull requests, which will show that the current PR cannot build for ARM microcontrollers (duplicate function definition).

**Comment by aykevl** [comment]
I did some measurements, and the changes in d6f7e54bdac6c3e73a6671420efa0f08f139f9a7...dcd49ef7c41505f5f8a5ec865a04a0ae8800b1fd caused a massive file size increase for WebAssembly targets. All tests increased significantly in size, with some of the worst offenders (in bytes) being binop.go (2747 -> 5740), channel.go (24884 -> 29216), float.go (4558 -> 7212), and init_multi.go (1043 -> 3702).

I'm afraid that increase in size is too much to be mergeable. It's not a problem if a refactor does increase output file size if it's small and the maintainability benefits are big, but such a huge increase in file size is simply not worth it.

On the other hand, the first commit d6f7e54bdac6c3e73a6671420efa0f08f139f9a7 only increases the channel.go test case in size by a somewhat significant size: 23022 -> 24884. But this is expected. The only other file size change I see is in reflect.go and is only a single byte so not a problem.
What do you think, maybe we can do just the buffered channel commit and look at refactoring the scheduler later?

**Comment by jadr2ddude** [comment]
@aykevl I think I might know why that is occurring, and I think I might be able to fix it. It most likely has something to do with my changes to deadlock. Unfortunately, the code LLVM was generating before was not necessarily correct. However, I should be able to come up with something that is both correct and efficient.

**Comment by jadr2ddude** [comment]
The growth in fully synchronous code has now been fixed. There are still a few things remaining to be fixed.

**Comment by deadprogram** [comment]
In addition to the other things you mention @jadr2ddude there are also now merge conflicts that need to be resolved.

**Comment by jadr2ddude** [comment]
Yes I know. I am actually trying an alternate approach now to see if it works better.

**Comment by jadr2ddude** [comment]
I am working on the merge conflicts now

**Comment by jadr2ddude** [comment]
So at this point, this is probably ready for review. There are a bunch of optimizations I want to do at some point, but I think this just needs to get in first.

**Comment by jadr2ddude** [comment]
Despite the failure of that one build I think this is working correctly. The `testdata/channel.go` test is just flaky.

**Comment by jadr2ddude** [comment]
I squashed this down to one commit.

**Comment by jadr2ddude** [comment]
fixes #526

**Comment by deadprogram** [comment]
@jadr2ddude this need the merge conflict  from https://github.com/tinygo-org/tinygo/pull/532 resolved now, if you please.

**Comment by jadr2ddude** [comment]
@aykevl the size regression should be fixed in 60bdf4fd18251158e9de80d69ad9dfcff9f5ea89

**Comment by aykevl** [comment]
I confirmed that the size increase on wasm is no longer present. However, I still see a rather significant increase in code size for code that doesn't need a scheduler (~450 bytes flash and ~90 bytes RAM) - in this case on the micro:bit. Is it _really_ necessary to enable the scheduler in all cases?
Code to test (fish shell):

```
for t in testdata/*.go; echo $t; tinygo build -o test.elf -target microbit -no-debug -size=short $t; end
```

testdata/binop.go size increase:

```
testdata/binop.go
   code    data     bss |   flash     ram
   2148       0    2196 |    2148    2196
testdata/binop.go
   code    data     bss |   flash     ram
   3252      56    2228 |    3308    2284
```

Remember, the whole purpose of TinyGo is small binaries. A lot of features would be simpler if this requirement would be dropped - for example reflect would be a whole lot easier to implement. So the benefits have to be big to allow such a size increase.

@deadprogram what's your view on this?

**Comment by jadr2ddude** [comment]
@aykevl When I fixed scheduler inclusion logic earlier, it seems that I did not update the logic for the tasks scheduler. That should be now fixed.

**Comment by aykevl** [comment]
Great, the size increase for cortex-m appears to be solved!

**Comment by aykevl** [comment]
I did some more size comparisons, and it turns out that this PR _reduces_ binary size in many cases (`make smoketest RISCV=0`). I'm not sure why, but it might need some extra testing in case it is caused by a bug (see #507 for example). I couldn't find any unexpected increases in size, so it's good on that front :smile:

**Comment by jadr2ddude** [comment]
@aykevl We do most certainty have correct size decreases in some cases. On WASM, I actually changed it so the scheduler is not always included. It is also entirely possible that some of this is easier for LLVM to optimize.

**Comment by deadprogram** [comment]
I tried to test this branch using the "apa102" example that is part of the drivers using the "Bluepill". It did not work as expected, although it does work in the `dev` branch. 

The issue is that LEDs lightup but do not cycle. I also tested with the ItsyBitsy-M4 and did not work as expected either, with the same issue..

I did merge in the latest SPI fix for STM32 into this branch before testing, so seems like something else wrong with the time on the STM32 possibly.

I also tested both the apa102 and the echo example using this branch on the ItsyBity-M4 and neither worked as expected.

I tested the "flightstick" example using the Arduino Nano33 and it worked perfectly using this branch, so it might be some platform specific or related to that specific example that allows it to work. I suspect that it has to do with the flightstick uses a go routine, and the other examples do not.

**Comment by jadr2ddude** [comment]
@deadprogram It appears that we hit a false positive, and that may not be using the scheduler even though it needs it. The large gap between the coroutines implementation and the tasks implementation seems to be a larger problem than I anticipated. . .

File: compiler/goroutine-lowering.go [hunk 1]
```diff
@@ -328,18 +340,349 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 				panic("expected const ptrtoint operand of runtime.makeGoroutine")
 			}
 			goroutine := ptrtoint.Operand(0)
+			if strings.HasPrefix(goroutine.Name(), "runtime.") {
+				continue
```
[the remark this answers is not in the document]
Comment (jadr2ddude) [reply compiler/goroutine-lowering.go]: So this is for `runtime.getFakeCoroutine`. If you would like, I can change it to just affect `runtime.fakeCoroutine`.

File: compiler/goroutine-lowering.go [hunk 2]
```diff
@@ -328,18 +340,349 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
...
+					if !broken && !inst.IsACallInst().IsNil() && inst.CalledValue() == yield {
+						coroDebugPrintln("broke", f.Name(), bb.AsValue().Name())
+						broken = true
+						c.builder.SetInsertPointBefore(inst)
+						c.createRuntimeCall("noret", []llvm.Value{}, "")
+						if f.Type().ElementType().ReturnType().TypeKind() == llvm.VoidTypeKind {
+							c.builder.CreateRetVoid()
+						} else {
+							c.builder.CreateRet(llvm.Undef(f.Type().ElementType().ReturnType()))
+						}
```
[the remark this answers is not in the document]
Comment (jadr2ddude) [reply compiler/goroutine-lowering.go]: This return is 100% reachable. What the code here does is it replaces the yield with a return that does not reactivate the caller. If all cases are known to behave like this, we propagate that information to the caller of this function. So no, an `unreachable` would be incorrect here.

File: compiler/goroutine-lowering.go [hunk 3]
```diff
@@ -328,18 +340,349 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
...
+
+						// insert return
+						c.builder.SetInsertPointBefore(next)
+						c.createRuntimeCall("noret", []llvm.Value{}, "")
+						var retInst llvm.Value
+						if f.Type().ElementType().ReturnType().TypeKind() == llvm.VoidTypeKind {
+							retInst = c.builder.CreateRetVoid()
+						} else {
+							retInst = c.builder.CreateRet(llvm.Undef(f.Type().ElementType().ReturnType()))
+						}
```
[the remark this answers is not in the document]
Comment (jadr2ddude) [reply compiler/goroutine-lowering.go]: Again, this is not what this code does.

File: compiler/goroutine-lowering.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aykevl) [reply compiler/goroutine-lowering.go]: Okay, thanks for the clarification. Makes sense.

File: compiler/goroutine-lowering.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aykevl) [reply compiler/goroutine-lowering.go]: Yes, that would be better. It would avoid surprising behavior in the runtime.

**Comment by aykevl** [comment]
I looked at the last commit, it looks good to me. I checked file sizes again, and there is nothing unexpected (only the known 500b flash, 56b RAM increase on Cortex-M, and obviously the channels test has grown bigger). So from my POV, nothing is blocking this PR anymore.

@deadprogram can you run your tests again, to make sure it breaks none of your code? This is a really large PR so there's a decent chance it breaks something.

**Comment by deadprogram** [comment]
I ran thru a group of tests on several different physical hardware devices, including code against a couple of different processors.

Everything worked exactly as it had previously.

Based on all that I am going to squash/merge this into `dev`.

Impressive work @jadr2ddude thank you for this contribution.

**Pull request merged**